internal/operator/snapshotlease/snapshotlease.go

```diff
@@ -0,0 +1,200 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package snapshotlease
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/gardener/etcd-druid/internal/common"
+	druiderr "github.com/gardener/etcd-druid/internal/errors"
+	"github.com/gardener/etcd-druid/internal/operator/component"
+	"github.com/gardener/etcd-druid/internal/utils"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	coordinationv1 "k8s.io/api/coordination/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ErrGetSnapshotLease indicates an error in getting the snapshot lease resources.
+	ErrGetSnapshotLease druidv1alpha1.ErrorCode = "ERR_GET_SNAPSHOT_LEASE"
+	// ErrSyncSnapshotLease indicates an error in syncing the snapshot lease resources.
+	ErrSyncSnapshotLease druidv1alpha1.ErrorCode = "ERR_SYNC_SNAPSHOT_LEASE"
+	// ErrDeleteSnapshotLease indicates an error in deleting the snapshot lease resources.
+	ErrDeleteSnapshotLease druidv1alpha1.ErrorCode = "ERR_DELETE_SNAPSHOT_LEASE"
+)
+
+type _resource struct {
+	client client.Client
+}
+
+// New returns a new snapshot lease operator.
+func New(client client.Client) component.Operator {
+	return &_resource{
+		client: client,
+	}
+}
+
+// GetExistingResourceNames returns the names of the existing snapshot leases for the given Etcd.
+func (r _resource) GetExistingResourceNames(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) ([]string, error) {
+	resourceNames := make([]string, 0, 2)
+	// We have to get snapshot leases one lease at a time and cannot use label-selector based listing
+	// because currently snapshot lease do not have proper labels on them. In this new code
+	// we will add the labels.
+	// TODO: Once all snapshot leases have a purpose label on them, then we can use List instead of individual Get calls.
+	deltaSnapshotObjectKey := client.ObjectKey{Name: etcd.GetDeltaSnapshotLeaseName(), Namespace: etcd.Namespace}
+	deltaSnapshotLease, err := r.getLease(ctx, deltaSnapshotObjectKey)
+	if err != nil {
```

Comment:
how do you propose to do that?
`deltaSnapshotLease` variable is used afterwards as well. I am assuming by inline you mean something like:
```go
if deltaSnapshotLease, err  := r.getLease(ctx, deltaSnapshotLeaseObjectKey); err != nil {
}
```
This will make deltaSnapshotLease local to the if condition.

Reply:
Apologies, my bad. I saw that it was being used afterwards as well but I overlooked the part that it was being initialised there itself. For some reason I thought that both the snapshot leases were initialised before.